### ISSUE TITLE
Clean _traits classes

### ISSUE THREAD

**Issue by PDoakORNL** [QMCPACK/qmcpack#3378]
There are many `_traits` classes in QMCPACK and while they attempt to make getting dependent type information and dependent relationships of type from template parameters or the application level ifdef'd type aliases they've become stale.

**Describe the solution you'd like**
Go through these classes 1 by 1, replace with new code capable of taking over the necessary work.

First target:
**scalar_traits.h should be removed.**
scalar_traits all of its functionality is redundant and most clashes with normal c++ idiom.
Familiar
`std::complex<T>::value_type = T;`
surprising
```c++
scalar_traits<std::complex<T>::value_type == std::complex<T>
scalar_traits<std::complex<T>::real_type == T
```

Type conversion via return arguments with many specialized versions of 
```
template<typename T1, typename T2>
void convert(const T1& in, T2& out);
```
These are inconsistent in their results.
- They do static_casts and assignment
- For complex and floating point types it calls .real() on the complex argument and assigns to the fp out.
- They provides a redundant version of the type templated TinyVector operator= 
- They provide a redundant version of OhmmsPete containers operator=
- They provide a dubious assignment between OhmmsPETE objects of unequal size with no protection against
overruns and semantics for Matrices and Tensors that are unexpected i.e. linearly into the raw data instead of into some corner of the object.

the scalar_traits themselves mostly serve to provide support to container_proxy which provides a seemingly unecessary way of wrapping Ohmms and STL objects containing a complex<T> type as a blob of T of twice the size.  

They are also used as a way to get Real type underlying a Value type.
```c++
typename scalar_traits<Value>::real_type my_real
```

Replace use with `complex_help.hpp` used in testing and batched code.
```c++
RealAlias<Value> my_real
```
`complex_help.hpp` also provides a IsComplex<T> and IsReal<T> appropriate for `if constexpr` and `std::enable_if_t` use.

**Describe alternatives you've considered**
Using these classes in new code instead of cleaner alternatives.

**Additional context**

**Comment by prckent** [comment]
I agree that these have become stale and are now more obfuscating than helpful, among other things. So cleaner versions, removal and replacement by modern C++, or simply deduplicating by using  infrastructure that we already have elsewhere would all be a help.

**Comment by ye-luo** [comment]
I think @PDoakORNL pointed out of an important implementation detail  that we messed up.
scalar_traits.h was totally misused under QMCWaveFunctions.
scalar_traits.h is intended to discover the basic plain-old type in a container. This is intended for interacting with external library like hdf5/mpi. We lack documentation there.
Complex is a container of size 2. For this reason, some complex to real logic is implemented there.

However, all the complex to real discovery is a separate concept and it should be implemented separately. scalar_traits should be a user of that implementation rather than a primary provider of that implementation.

I think having RealAlias is a good move. @PDoakORNL Is it OK to take RealAlias out of #3362 and replace all the misuse of scalar_traits. Also move all the real/imag/conj out of `scalar_traits.h` to `complex_helper.hpp`. I have comments on the code which is already in #3362 for this purpose. If separated, we can work on two topics separately in parallel. #3362 doesn't need to updated immediately, I can review it by ignoring all the bits related to this issue topic.

**Comment by ye-luo** [comment]
I changed my mind. All the issues needs to be fixed but after #3362.
